Documentation/design-docs/COM-activation.md

```diff
@@ -0,0 +1,139 @@
+
+During .NET class registration, the shim is identified as the in-proc server for the class. Additional metadata is inserted into the registry to indicate what .NET assembly to load and what type to activate. For example, in addition to the typical [in-proc server](https://docs.microsoft.com/en-us/windows/desktop/com/inprocserver32) registry values the following values are added to the registry for the `TypeLoadException` class.
+
+```
+"Assembly"="mscorlib, Version=1.0.5000.0, Culture=neutral, PublicKeyToken=b77a5c561934e089"
+"Class"="System.TypeLoadException"
+"RuntimeVersion"="v1.1.4322"
+```
+
+The above registration is typically done with the [`RegAsm.exe`](https://docs.microsoft.com/en-us/dotnet/framework/tools/regasm-exe-assembly-registration-tool) tool. Alternatively, registry scripts can be generated by `RegAsm.exe`.
+
+### .NET Core Class COM Activation
+
+In .NET Core, our intent will be to avoid a system wide shim library. This decision may add additional cost for deployment scenarios, but will reduce servicing and engineering costs by making deployment more explicit and less magic.
+
+The current .NET Core hosting solutions are described in detail at [Documentation/design-docs/host-components.md](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/host-components.md). Along with the existing hosts an additional COM activation host library will be added. This library will export the required functions for COM class activation and act in a way similar to `mscoree.dll`.
+
+>[`HRESULT DllGetClassObject(REFCLSID rclsid, REFIID riid, LPVOID *ppv);`](https://docs.microsoft.com/en-us/windows/desktop/api/combaseapi/nf-combaseapi-dllgetclassobject)
+
+>[`HRESULT DllCanUnloadNow();`](https://docs.microsoft.com/en-us/windows/desktop/api/combaseapi/nf-combaseapi-dllcanunloadnow)
+
+When `DllGetClassObject()` is called in an activation scenario, the following will occur:
+
+1) Determine additional registration information needed for activation.
+    * **The location and mechanics of accessing this registration information is still TBD**
+    * `Assembly` (**required**) - the [Fully-Qualified Name](https://docs.microsoft.com/en-us/dotnet/framework/reflection-and-codedom/specifying-fully-qualified-type-names) of the assembly.
+    * `Class` (**required**) - the full type name (e.g. `MyCompany.MyProduct.MyClass`).
+    * `Codebase` (**required**) - an absolute path to the assembly to load. If a [`runtimeconfig.json`](https://github.com/dotnet/cli/blob/master/Documentation/specs/runtime-configuration-file.md) file exists adjacent to the assembly that file will be preferred to describe CLR configuration details.
+    * `RuntimeVersion` (optional) - CLR version requirement.
+1) Using the existing `hostfxr` library, attempt to discover the desired CLR.
```

Comment:
This is actually more than just CLR. We need to have the right frameworks as well. In fact it goes hand in hand. Currently without multiple frameworks the runtime config may specify for example ASP.NET framework, which than depends on the NETCore.App which is the one actually identifying the runtime to use.
In the future with multiple framework support this gets even more complicated.
Currently I'm not aware of a plan to support loading multiple versions of higher-level frameworks (like ASP.NET) side-by-side into the same process (even if they depend on the same underlying NETCore.App). So the resolution logic here will have to consider all of that.
That said, it already does so, so the implementation is probably almost ready, but I think it's worth noting it here, that this is not just about the runtime version, but also about all the frameworks.

Reply:
@vitek-karas The concept of different frameworks is something I am still working on fully incorporating in my thought process - thanks for keeping me honest.

Since the framework is also a consideration, it seems that the `deps.json` file is going to be a requirement, where as the `runtimeconfig.json` could be optional? Would that address your concerns here?

Reply:
I'm pretty sure frameworks are specified in `runtimeconfig.json`. `deps.json` can only customize the frameworks a little (override assembly at a time), but usually it just does nothing to them. So this would make `runtimeconfig.json` basically mandatory. It might be possible to not have `deps.json`... but I'm not sure.